sample-programs/erc721/erc721.scm

```diff
@@ -0,0 +1,131 @@
+      (get spender 
+        (fetch-entry tokens-spender (tuple (token-id token-id))))))
+    (if (eq? spender 'null) 'false 'true)))
+
+;; Tells whether an operator is approved by a given owner (isApprovedForAll method in ERC721)
+(define (is-operator-approved (account principal) (operator principal))
+  (let ((is-approved
+      (get is-approved 
+        (fetch-entry accounts-operator (tuple (operator operator) (account account))))))
+    (if (eq? is-approved 'null) 'false is-approved)))
+
+;; Returns whether the given actor can transfer a given token ID.
+;; To be optimized
+(define (can-transfer (actor principal) (token-id int)) 
+  (or 
+    (eq? actor (owner-of token-id)) 
+    (is-spender-approved actor token-id)
+    (is-operator-approved (owner-of token-id) actor)))
+ 
+;; Internal - Register token
+(define (register-token! (new-owner principal) (token-id int))
+  (let ((current-balance (balance-of new-owner)))
+    (begin
+      (set-entry! tokens-owner 
+        (tuple (token-id token-id))
+        (tuple (owner new-owner))) 
+      (set-entry! tokens-count 
+        (tuple (owner new-owner))
+        (tuple (count (+ 1 current-balance)))) 
+      'true)))
+
+;; Internal - Release token
+(define (release-token! (owner principal) (token-id int))
+  (let ((current-balance (balance-of owner)))
+    (begin
+      (delete-entry! tokens-spender 
+        (tuple (token-id token-id))) 
+      (set-entry! tokens-count 
+        (tuple (owner owner))
+        (tuple (count (- current-balance 1)))) 
+      'true)))
+
+;; Public functions
+
+;; Approves another address to transfer the given token ID (approve method in ERC721)
+;; To be optimized
+(define-public (set-spender-approval (spender principal) (token-id int))
+  (if (eq? spender tx-sender)
+    'false
+    (if (or (eq? tx-sender (owner-of token-id)) 
+            (is-operator-approved tx-sender (owner-of token-id)))
+      (begin
+        (set-entry! tokens-spender 
+          (tuple (token-id token-id))
+          (tuple (spender spender))) 
+      'true)
+      'false)))
+
+;; Sets or unsets the approval of a given operator (setApprovalForAll method in ERC721)
+(define-public (set-operator-approval (operator principal) (is-approved bool))
+  (if (eq? operator tx-sender)
+    'false
+    (begin
+      (set-entry! accounts-operator 
+        (tuple (operator operator) (account tx-sender))
+        (tuple (is-approved is-approved))) 
+    'true)))
+
+;; Transfers the ownership of a given token ID to another address.
+(define-public (transfer-from (owner principal) (recipient principal) (token-id int))
+  (if (can-transfer tx-sender token-id)
+    (and
+      (release-token! owner token-id)
+      (register-token! recipient token-id))
+    'false))
+
+;; Transfers tokens to a specified principal.
+(define-public (transfer (recipient principal) (token-id int))
+  (transfer-from tx-sender recipient token-id))
+
+;; Mint new tokens.
+(define-public (mint! (owner principal) (token-id int))
```

Comment:
Is anyone allowed to make new tokens unconditionally?

Reply:
Related to https://github.com/blockstack/blockstack-core/pull/955#discussion_r285348131

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/955/commits/542070dbebf75db31650003d175a36fccd615e98